android/sdl_android/src/main/java/com/smartdevicelink/transport/RouterServiceValidator.java

```diff
@@ -200,6 +214,187 @@ public boolean validate(){
+			if(wasInstalledByAppStore(packageName)){ //Was this package installed from a trusted app store
+				if( isTrustedPackage(packageName, pm)){//Is this package on the list of trusted apps.
+					valid = true;
+				}
+			}
+		}
+		if (callback != null) {
+			callback.onFinishedValidation(valid, component);
+			if (valid) {
+				synchronized (this) {
+					this.service = component;
+				}
+			}
+		}
+	}
+	/**
+	 * This method retrieves the best routerservice name asynchronously.
+	 * @param context
+	 */
+	//private void retrieveBestRouterServiceName(Context context) {
+	//	FindRouterTask task = new FindRouterTask(null);
+	//	task.execute(context);
+	//}
+
+	/**
+	 * FindRouterTask: AsyncTask to find the connected RouterService.
+	 */
+	class FindRouterTask extends AsyncTask<Context, Void, ComponentName> {
+		FindConnectedRouterCallback mCallback;
+		final Handler mHandler = new Handler(Looper.getMainLooper());
+		final Integer TIMEOUT_MSEC = 10000; // 10 sec
+
+		FindRouterTask(FindConnectedRouterCallback callback) {
+			mCallback = callback;
+		}
+
+		@Override
+		protected ComponentName doInBackground(final Context... contexts) {
+			// let's use ServiceFinder here
+			final BlockingQueue<ComponentName> serviceQueue = new LinkedBlockingQueue<>();
+			final AtomicInteger _counter = new AtomicInteger(0);
+			Context context = contexts[0];
+			final Thread _currentThread = Thread.currentThread();
+			new ServiceFinder(context, context.getPackageName(), new ServiceFinder.ServiceFinderCallback() {
+				@Override
+				public void onComplete(Vector<ComponentName> routerServices) {
+					// OK, we found the routerServices. Let's see one-by-one.
+					if (routerServices == null || routerServices.isEmpty()) {
+						_currentThread.interrupt();
+						return;
+					}
+
+
+					final int numServices = routerServices.size();
+					for (ComponentName name: routerServices) {
+						final SdlRouterStatusProvider provider = new SdlRouterStatusProvider(contexts[0], name, new SdlRouterStatusProvider.ConnectedStatusCallback() {
+							@Override
+							public void onConnectionStatusUpdate(final boolean connected, final ComponentName service, final Context context) {
+								// make sure this part runs on main thread.
+								mHandler.post(new Runnable() {
+									@Override
+									public void run() {
+										_counter.incrementAndGet();
+										if (connected) {
+											DebugTool.logInfo("We found the connected service (" + service + "); currentThread is " + Thread.currentThread().getName());
+											serviceQueue.add(service);
+										} else if (_counter.get() == numServices) {
+											DebugTool.logInfo("SdlRouterStatusProvider returns service=" + service + "; connected=" + connected);
+											_currentThread.interrupt();
+										}
+									}
+								});
+							}
+						});
+						DebugTool.logInfo("about checkIsConnected; thread=" + Thread.currentThread().getName());
+						provider.checkIsConnected();
```

Comment:
Would it be possible to modify the `SdlRouterStatusProvider.bindToService()` to call `SdlBroadcastReceiver.setForegroundExceptionHandler()` before calling `startForegroundService()` similar to how we do it here:

https://github.com/smartdevicelink/sdl_java_suite/blob/797b7a14a7fd67fdf6faa504865e4e22228858ef/android/sdl_android/src/main/java/com/smartdevicelink/transport/SdlBroadcastReceiver.java#L257

This is just to prevent an exception from happening in the rare case where the router service doesn't enter foreground on time. It is okay if the `setForegroundExceptionHandler()`'s modifier is changed to protected.

Reply:
That makes sense. @bilal-alsharifi, thanks for the suggestion. Updated the PR accordingly.